research/minigo/minigo.py

```diff
@@ -0,0 +1,406 @@
+    estimator_working_dir: tf.estimator working directory.
+    training_chunk_dir: Directory where gathered training chunks are.
+    params: Hyper params of the model.
+  """
+  model_num, model_name = model_utils.get_latest_model(models_dir)
+  print('Initializing from model %s' % model_name)
+
+  new_model_name = model_utils.generate(model_num + 1)
+  print('New model will be %s' % new_model_name)
+  save_file = os.path.join(models_dir, new_model_name)
+
+  try:
+    tf_records = sorted(
+        tf.gfile.Glob(os.path.join(training_chunk_dir, '*.tfrecord.zz')))
+    tf_records = tf_records[
+        -1 * (params.train_window_size // params.examples_per_chunk):]
+
+    print('Training from: %s to %s' % (tf_records[0], tf_records[-1]))
+
+    with utils.logged_timer('Training'):
+      dual_net.train(estimator_working_dir, tf_records, model_num + 1, params)
+      dual_net.export_model(estimator_working_dir, save_file)
+  except tf.errors.UnknownError:
+    print('Got an error training, muddling on...')
+    tf.logging.error('Train error')
+
+
+def validate(models_dir, holdout_dir, estimator_working_dir, params):
+  """Validate the latest model on the holdout dataset.
+
+  Args:
+    models_dir: Directories where the completed generations/models are.
+    holdout_dir: Directories where holdout data are.
+    estimator_working_dir: tf.estimator working directory.
+    params: Hyperparameters.
+  """
+  model_num, model_name = model_utils.get_latest_model(models_dir)
+
+  # Get the holdout game data
+  nums_names = model_utils.get_models(models_dir)
+  models = [num_name for num_name in nums_names if num_name[0] < model_num]
+
+  holdout_dirs = [os.path.join(holdout_dir, pair[1])
+                  for pair in models[-params.holdout_generation:]]
+  checkpoint_name = os.path.join(models_dir, model_name)
+
+  tf_records = []
+  with utils.logged_timer('Building lists of holdout files'):
+    for record_dir in holdout_dirs:
+      if os.path.exists(record_dir):  # make sure holdout dir exists
+        tf_records.extend(tf.gfile.Glob(os.path.join(record_dir, '*.zz')))
+
+  print('The length of tf_records is %d.' % len(tf_records))
+  with utils.logged_timer('Validating from %s to %s' % (
+      os.path.basename(tf_records[0]), os.path.basename(tf_records[-1]))):
+    dual_net.validate(estimator_working_dir, tf_records,
+                      checkpoint_name, params)
+
+
+def evaluate(models_dir, black_model_name, white_model_name,
+             evaluate_dir, params):
+  """Evaluate on two models.
+
+  Args:
+    models_dir: Directories where the completed generations/models are.
+    black_model_name: The name of the model playing black
+    white_model_name: The name of the model playing white
+    evaluate_dir: Where to write the evaluation results'='sgf/evaluate/
+    params: Hyperparameters.
+
+  Returns:
+    The model name of the winner.
+  """
+
+  black_model = os.path.join(models_dir, black_model_name)
```

Comment:
Again, caching the models is very useful. We should either allow the models to be passed in, or we could add a #games parameters and loop here.